[Clang] Add constant evaluation support for x86 psadbw

Mohxen opened the pull request:
This is the first draft of adding constant evaluation support for the 
x86 PSADBW and PSADBW128 intrinsics in Clang’s constexpr interpreter.

The patch currently:
- Adds initial handling logic in ExprConstant.cpp
- Includes an early version of the test under 
  clang/test/AST/ByteCode/x86-psadbw-psadbw128.cpp

This PR is still under development. Feedback on structure, integration 
location, and testing is welcome before polishing the final version.

Fixes #188062

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -12078,6 +12080,54 @@ static bool evalPackBuiltin(const CallExpr *E, EvalInfo &Info, APValue &Result,
+    uint64_t a = A.getVectorElt(i).getInt().getZExtValue();
+    uint64_t b = B.getVectorElt(i).getInt().getZExtValue();
+    Sum1 += (a > b ? a - b : b - a);
+  }
```
Convert to a format that will handle 256/512 variants now - otherwise you'll just end up rewriting all of this later on.
```
SmallVector<APValue, 8> Elts;
for (unsigned Lane = 0; Lane != Len; Lane += 8) {
  for (unsigend I = 0; I != 8; ++I) {
    APInt A = A.getVectorElt(Lane + I).getInt();
    APInt B = B.getVectorElt(Lane + I).getInt();
  }
etc.
  Elts.emplace_back(APValue(APSInt(APInt(64, Sum), Unsigned)));
}
```

RKSimon commented:
@Mohxen ping?

RKSimon commented:
@Mohxen ping?

Mohxen committed:
[Clang] Add constexpr support for x86 PSADBW intrinsics

Mohxen committed:
[Clang] Add constexpr support for x86 PSADBW intrinsics

Mohxen commented:
Hi @RKSimon,

I updated this draft based on your feedback. This now covers the PSADBW intrinsics requested in #188062. This was also split out from #157522.

The patch now handles `psadbw128`, `psadbw256`, and `psadbw512` in both constexpr evaluators, uses `APIntOps::abdu` for the unsigned absolute difference, and keeps the fixed-vector checks as asserts.

I also moved the builtins into existing `Constexpr` builtin groups, marked the public wrappers constexpr, and replaced the separate AST/ByteCode test with `TEST_CONSTEXPR` checks in the existing X86 builtin test files.

Covered wrappers:
`_mm_sad_pu8`, `_mm_sad_epu8`, `_mm256_sad_epu8`, and `_mm512_sad_epu8`.

Local targeted tests pass: `mmx-builtins.c`, `sse2-builtins.c`, `avx2-builtins.c`, and `avx512bw-builtins.c`.

Please let me know your feedback.

RKSimon commented on clang/include/clang/Basic/BuiltinsX86.td:
```diff
@@ -173,13 +173,13 @@ let Features = "sse2", Attributes = [NoThrow] in {
 let Features = "sse2", Attributes = [NoThrow, Const, Constexpr, RequiredVectorWidth<128>] in {
   def cvtpd2ps : X86Builtin<"_Vector<4, float>(_Vector<2, double>)">;
   def cvtsd2ss : X86Builtin<"_Vector<4, float>(_Vector<4, float>, _Vector<2, double>)">;
+  def psadbw128 : X86Builtin<"_Vector<2, long long int>(_Vector<16, char>, _Vector<16, char>)">;
```
we should probably update the signature to:
```suggestion
  def psadbw128 : X86Builtin<"_Vector<2, long long int>(_Vector<16, unsigned char>, _Vector<16, unsigned char>)">;
```
same for the others (might require changes to the casts in the headers)

Mohxen committed:
Address PSADBW unsigned byte builtin signatures

Mohxen committed:
Address PSADBW unsigned byte builtin signatures

RKSimon reviewed (changes requested):
@Mohxen any update on those CI failures?

Mohxen committed:
[clang][X86] Fix psadbw builtin smoke test argument type

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into feature-psadbw-constexpr
